_drafts/zero-to-main-libc.md

```diff
@@ -0,0 +1,385 @@
+<!-- excerpt end -->
+
+## Setup
+As we did in our previous posts, we are using Adafruit’s Metro M0 development board to run our examples. We use a cheap CMSIS-DAP adapter and openOCD to program it.
+
+You can find a step by step guide [in our previous post](https://interrupt.memfault.com/blog/getting-started-with-ibdap-and-atsamd21g18).
+
+As with previous examples, we start with our “minimal” example which you can find [on GitHub](https://github.com/memfault/zero-to-main/tree/master/minimal). I’ve reproduce the source code for `main.c` below:
+
+```c
+#include <samd21g18a.h>
+
+#include <port.h>
+#include <string.h>
+
+#define LED_0_PIN PIN_PA17
+
+static void set_output(const uint8_t pin) {
+  struct port_config config_port_pin;
+  port_get_config_defaults(&config_port_pin);
+  config_port_pin.direction = PORT_PIN_DIR_OUTPUT;
+  port_pin_set_config(pin, &config_port_pin);
+  port_pin_set_output_level(pin, false);
+}
+
+int main() {
+  memcpy(NULL, NULL, 0);
+  set_output(LED_0_PIN);
+  while (true) {
+    port_pin_toggle_output_level(LED_0_PIN);
+    for (int i = 0; i < 100000; ++i) {}
+  }
+}
+``` 
+
+## Implementing Newlib
+### Why Newlib?
+
+There are several implementations of the C Standard Library, starting with the venerable `glibc` found on most GNU/Linux systems. Alternative implementations include Musl libc, Bionic libc, ucLibc, and dietlibc. 
+
+Newlib is an implementation of the C Standard Library targeted at bare-metal embedded systems that is maintained by RedHat.
+
+Newlib has become the de-facto standard in embedded software because it is complete, has optimizations for a wide range of architectures, and produces relatively small code.
+
+Today it is bundled alongside toolchains and SDK provided by vendors such as ARM (`arm-none-eabi-gcc`) and Espressif (ESP-IDF for ESP32).
+
+> **newlib vs. newlib-nano**: when code-size constrained, you may chooses to use a variant of newlib, called newlib-nano, which does away with some C99 features, and some `printf` bells and whistles to deliver a more compact standard library. Newlib-nano is enabled with the `—specs=nano.specs` CFLAG. You can read more about it in our [code size blog post](https://interrupt.memfault.com/blog/code-size-optimization-gcc-flags#c-library)  
+
+### Enabling Newlib
+
+Newlib is enabled **by default** when you build a project with `arm-none-eabi-gcc`. Indeed, you must explicitly opt-out with `-nostdlib` if you prefer to build your firmware without it.
+
+This is what we do for our “minimal” example, to guarantee we do not include any libc functionality by mistake.
+
+```makefile
+PROJECT := minimal
+BUILD_DIR ?= build
+
+CFLAGS += -nostdlib
+
+SRCS = \
+	startup_samd21.c \
+	$(PROJECT).c
+
+include ../common-standalone.mk
+```
+
+It is very easy to add a dependency on the C standard library without meaning to, as GCC will sometimes use standard C functions implicitly. For example, consider this code used to zero-initialize a struct:
+
+```c
```

Comment:
Ooo, awesome example!

Nit: I'd probably use  a slightly bigger buffer in the example, below a certain size the memset will usually be unrolled ...

```
extern void noop(void *p);
__attribute__((noinline))
void example_memset(void) {
    int b[3] = { 0 };
    noop(b);
}

// int b[3] = { 0 };
b4:       b51f            push    {r0, r1, r2, r3, r4, lr}
b6:       2300            movs    r3, #0
b8:       a801            add     r0, sp, #4
ba:       e9cd 3301       strd    r3, r3, [sp, #4]
be:       9303            str     r3, [sp, #12]
c0:       f000 f863       bl      18a <noop>
c4:       b005            add     sp, #20
c6:       f85d fb04       ldr.w   pc, [sp], #4


// int b[50] = { 0 };
b4:       b500            push    {lr}
b6:       b0b3            sub     sp, #204        ; 0xcc
b8:       22c8            movs    r2, #200        ; 0xc8
ba:       2100            movs    r1, #0
bc:       4668            mov     r0, sp
be:       f001 fe63       bl      1d88 <memset>
c2:       4668            mov     r0, sp
c4:       f000 f863       bl      18e <noop>
c8:       b033            add     sp, #204        ; 0xcc
ca:       f85d fb04       ldr.w   pc, [sp], #4
```

Aside: I thought you could convince the compiler to not use any libc calls with some combination of fno-builtin, -ffreestanding, etc but I couldn't actually get the b[50] example to compile. Always hit:

```
/Users/chrisc/arm_8_2019q3/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: /Users/chrisc/dev/interrupt/example/freertos-example/build/main.o: in function `example_memset':
/Users/chrisc/dev/interrupt/example/freertos-example/main.c:57: undefined reference to `memset'
collect2: error: ld returned 1 exit status
```

Reply:
👍